File: GCDUserInterface/Project/ProjectTreeNode.vb
Namespace Project

    Public Class ProjectTreeNode

        Private m_Type As ucProjectExplorer.GCDNodeTypes
        Private m_Item As Object

        Public ReadOnly Property Item As Object
            Get
                Return m_Item
            End Get
        End Property

        Public ReadOnly Property NodeType As ucProjectExplorer.GCDNodeTypes
            Get
                Return m_Type
            End Get
        End Property

        Public Overrides Function ToString() As String
            Return Item.ToString()
        End Function

        Public Sub New(type As ucProjectExplorer.GCDNodeTypes, theItem As Object)
            m_Type = type
            m_Item = theItem
        End Sub

        Public Shared Operator =(node1 As ProjectTreeNode, node2 As ProjectTreeNode) As Boolean

            If node1 Is Nothing AndAlso node2 Is Nothing Then
                Return True
            ElseIf node1 Is Nothing Then
                Return False
            ElseIf node2 Is Nothing Then
                Return False
            Else
                If (node1.NodeType = node2.NodeType) Then
                    Return node1.Item Is node2.Item
                Else
                    Return False
                End If
            End If
        End Operator

        Public Shared Operator <>(node1 As ProjectTreeNode, node2 As ProjectTreeNode) As Boolean
            Return Not node1 = node2
        End Operator

    End Class

End Namespace

File: GCDCore/UserInterface/Project/ProjectTreeNode.cs
namespace GCDCore.UserInterface.Project
{
    public class ProjectTreeNode
    {
        public enum GCDNodeTypes
        {
            Project,
            InputsGroup,
            SurveysGroup,
            DEMSurvey,
            AssocGroup,
            AssociatedSurface,
            ErrorSurfaceGroup,
            ErrorSurface,
            ReferenceSurfaceGroup,
            ReferenceSurface,
            MasksGroup,
            Mask,
            AOIGroup,
            AOI,
            AnalysesGroup,
            ChangeDetectionGroup,
            ChangeDetectionDEMPair,
            DoD,
            BudgetSegregationGroup,
            BudgetSegregationMask,
            BudgetSegregation,
            MorphologicalAnalysisGroup,
            MorphologicalAnalysis,
            InterComparisonGroup,
            InterComparison,
            ReservoirGroup,
            Reservoir
        }

        private GCDNodeTypes m_Type;

        private object m_Item;
        public object Item
        {
            get { return m_Item; }
        }

        public GCDNodeTypes NodeType
        {
            get { return m_Type; }
        }

        public override string ToString()
        {
            return Item.ToString();
        }

        public string Name
        {
            get
            {
                string name = NodeType.ToString();
                if (Item != null)
                {
                    name = string.Format("{0}_{1}", name.ToString());
                }

                return name;
            }
        }

        public ProjectTreeNode(GCDNodeTypes type, object theItem)
        {
            m_Type = type;
            m_Item = theItem;
        }

        public bool Equals(ProjectTreeNode obj)
        {
            if (obj == null)
            {
                return false;
            }
            else
            {
                if (NodeType == obj.NodeType)
                {
                    return object.ReferenceEquals(Item, obj.Item);
                }
                else
                {
                    return false;
                }
            }
        }        
    }
}